File: jsrc/machine/SxxMachine/ListTerm.java
package com.googlecode.prolog_cafe.lang;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.IdentityHashMap;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

import com.googlecode.prolog_cafe.lang.Term.TermTreeIterator;
/**
 * List.<br>
 * The class <code>ListTerm</code> represents a list structure.<br>
 * 
 * <pre>
 *  % [1,2]
 *  Term Nil = SymbolTerm.makeSymbol("[]");
 *  Term  n1 = IntegerTerm(1);
 *  Term  n2 = IntegerTerm(2);
 *  Term   t = new ListTerm(n1, new ListTerm(n2, Nil));
 *  
 *  Term car = ((ListTerm)t).car();
 *  Term cdr = ((ListTerm)t).cdr();
 * </pre>
 *
 * Here is sample program for creating a list from <code>1</code> to <code>n</code>.
 * <pre>
 * public static Term makeList(int n) {
 *   Term t = SymbolTerm.makeSymbol("[]");
 *   for (int i=n; i>0; i--) {
 *     t = new ListTerm(new IntegerTerm(i), t);
 *   }
 *   return t;
 * }
 * </pre>
 *
 * @author Mutsunori Banbara (banbara@kobe-u.ac.jp)
 * @author Naoyuki Tamura (tamura@kobe-u.ac.jp)
 * @version 1.0
 */
public class ListTerm extends Term {
    /** A functor <code>'.' /2</code>. */
    protected static final SymbolTerm SYM_DOT = SymbolTerm.intern(".", 2);

    private final boolean immutable;

    /** Holds the first element of this <code>ListTerm</code>. */
    private final Term car;

    /**
     * Holds the list consisting of all the rest of the elements of 
     * this <code>ListTerm</code> but the first one.
     */
    private final Term cdr;

    /**
     * Constructs a new Prolog list structure
     * such that <code>_car</code> is the first element of this list, and 
     * <code>_cdr</code> is the list consisting of all the rest of the 
     * elements of this list but the first one.
     */
    public ListTerm(Term _car, Term _cdr) { 
		// TODO assert _car!=null;
    	// TODO assert _cdr!=null;
    	car = _car;
		cdr = _cdr;
		immutable = car.isImmutable() && cdr.isImmutable();
    }

    /** Returns the value of <code>car</code>.
     * @see #car
     */
    public final Term car() { return car; }

    /** Returns the value of <code>cdr</code>.
     * @see #cdr
     */
    public Term cdr() { return cdr; }

    /* Term */
    public final boolean unify(Term t, Trail trail) {
		Term p = this;
		t = t.dereference();
		while ((t instanceof ListTerm) && (p instanceof ListTerm)
				&& ((ListTerm) p).car.unify(((ListTerm) t).car, trail)) {
			p = ((ListTerm) p).cdr().dereference();
			t = ((ListTerm) t).cdr().dereference();
		}
		if (t instanceof VariableTerm) {
			return ((VariableTerm) t).bind(p, trail);
		}
		if (p instanceof VariableTerm) {
			return ((VariableTerm) p).bind(t, trail);
		}
		return !(t instanceof ListTerm) && !(p instanceof ListTerm) && p.unify(t, trail);
	}
    
    
    /** 
     * @return the <code>boolean</code> whose value is
     * <code>convertible(List.class, type)</code>.
     * @see Term#convertible(Class, Class)
     */
    public boolean convertible(Class type) { 
    	return convertible(List.class, type); 
    }

    protected Term copy(IdentityHashMap<VariableTerm,VariableTerm> copyHash) { 
    	if (immutable){
    		return this;
    	}
    	Deque<ListTerm> stack = new ArrayDeque<ListTerm>();
    	Term p = this;
    	while (p instanceof ListTerm && !((ListTerm)p).immutable){
    		ListTerm lt = (ListTerm) p;
    		stack.push(lt);
    		p = lt.cdr().dereference();
    	}
    	p = p.copy(copyHash);
    	while (!stack.isEmpty()){
    		ListTerm lt = stack.pop();
			p = new ListTerm(lt.car.copy(copyHash), p);
    	}
    	return p;
    }

    public boolean isGround() {
		return car.isGround() && cdr().isGround();
    }

    public final String name() { return SYM_DOT.name(); }

    public Term arg(int nth) {
      Term t = this;
      int old_nth = nth;
      while ((t instanceof ListTerm) && 0 < nth) {
        nth--;
        t = ((ListTerm)t).cdr().dereference();
      }
      if ((t instanceof ListTerm))
        return ((ListTerm)t).car;
      throw new ArrayIndexOutOfBoundsException(old_nth);
    }

    /** Returns the length of this <code>ListTerm</code>. */
    public int length() {
	int count = 0;
	Term t = this;
	while((t instanceof ListTerm)) {
	    count++;
	    t = ((ListTerm)t).cdr().dereference();
	}
	return count;
    }

    /** 
     * Returns a {@code java.util.List} corresponds to this <code>ListTerm</code>
     * according to <em>Prolog Cafe interoperability with Java</em>.
     * @return a {@link java.util.List} object equivalent to
     * this <code>IntegerTerm</code>.
     */
    public List toJava() { 
		List<Object> vec = new ArrayList<Object>();
		Term t = this;
		while((t instanceof ListTerm)) {
		    vec.add(((ListTerm)t).car.dereference().toJava());
		    t = ((ListTerm)t).cdr().dereference();
		}
		if (!t.isNil()){
			vec.add(t);
		}
		return vec;	
    }

    public void toQuotedString(StringBuilder sb){
		TermTreeIterator it = new TermTreeIterator(this);
		while(it.hasNext()){
			it.next().toQuotedString(sb);
		}
    }

    /* Object */
    /**
     * Checks <em>term equality</em> of two terms.
     * The result is <code>true</code> if and only if the argument is an instance of
     * <code>ListTerm</code>, and 
     * all corresponding pairs of elements in the two lists are <em>term-equal</em>.
     * @param obj the object to compare with. This must be dereferenced.
     * @return <code>true</code> if the given object represents a Prolog list
     * equivalent to this <code>ListTerm</code>, false otherwise.
     * @see #compareTo
     */
    public boolean equals(Object obj) {
		return obj instanceof ListTerm && car.equals(((ListTerm) obj).car().dereference()) && cdr().equals(((ListTerm) obj).cdr().dereference());
	}

    public int hashCode() {
	int h = 1;
	h = 31*h + SYM_DOT.hashCode();
	h = 31*h + car.dereference().hashCode();
	h = 31*h + cdr().dereference().hashCode();
	return h;
    }

    /** Adds a string representation of this <code>ListTerm</code> to given StringBuilder instance.
     * Non recursive implementation */
    public void toString(StringBuilder sb){
		TermTreeIterator it = new TermTreeIterator(this);
		while(it.hasNext()){
			it.next().toString(sb);
		}
    }

    @Override
    public Iterator<Term> iterator() {
    	return new ListTermIterator(this);
    }
    
    /**
     * Iterator over terms that make up this ListTerm.
     */
    private static class ListTermIterator implements Iterator<Term> {    	
    	private static final SymbolTerm LEFT_BRACKET = SymbolTerm.intern("[");
		private static final SymbolTerm SEPARATOR = SymbolTerm.intern("|");
		private static final SymbolTerm RIGHT_BRACKET = SymbolTerm.intern("]");
		private static final SymbolTerm COMMA = SymbolTerm.intern(",");
		
		
		private Term current;
    	private int index = 0; 
    	private Term[] tail = null;
    	private boolean comma = false;
    	private boolean first = true;
    	
    	
		public ListTermIterator(Term start) {
			this.current = start;
		}

		@Override
		public boolean hasNext() {
			return current!=null && (tail==null || index<tail.length);
		}

		@Override
		public Term next() {
			Term result;
			if (first) {
				first = false;
				return LEFT_BRACKET;
			} else if (comma){
				comma = false;
				return COMMA;
			} else if (current instanceof ListTerm) {
				result = ((ListTerm)current).car;
				current = ((ListTerm)current).cdr().dereference();
				comma = (current instanceof ListTerm);
			} else if (current.isNil()){
				result = RIGHT_BRACKET;
				current = null;
			} else if (tail==null){
				tail = new Term[]{
					current,
					RIGHT_BRACKET
				};
				result = SEPARATOR;
				index = 0;
			} else if (index<tail.length){
				result = tail[index];
				index++;
			} else {
				throw new NoSuchElementException();
			}
			return result;
		}
    }

    /* Comparable */
    /** 
     * Compares two terms in <em>Prolog standard order of terms</em>.<br>
     * It is noted that <code>t1.compareTo(t2) == 0</code> has the same
     * <code>boolean</code> value as <code>t1.equals(t2)</code>.
     * @param anotherTerm the term to compared with. It must be dereferenced.
     * @return the value <code>0</code> if two terms are identical; 
     * a value less than <code>0</code> if this term is <em>before</em> the <code>anotherTerm</code>;
     * and a value greater than <code>0</code> if this term is <em>after</em> the <code>anotherTerm</code>.
     */
    public int compareTo(Term anotherTerm) { // anotherTerm must be dereferenced.
	if ((anotherTerm instanceof VariableTerm) || (anotherTerm instanceof NumberTerm) || (anotherTerm instanceof SymbolTerm))
	    return AFTER;
	if ((anotherTerm instanceof StructureTerm)) {
	    int arity = anotherTerm.arity();
	    if (2 != arity)
		return (2 - arity);
	    SymbolTerm functor = ((StructureTerm)anotherTerm).functor();
	    if (! SYM_DOT.equals(functor))
		return SYM_DOT.compareTo(functor);
	}
	Term[] args = new Term[2];
	if ((anotherTerm instanceof ListTerm)) {
	    args[0] = ((ListTerm)anotherTerm).car();
	    args[1] = ((ListTerm)anotherTerm).cdr();
	} else if ((anotherTerm instanceof StructureTerm)) {
	    args = ((StructureTerm)anotherTerm).args();
	} else {
	    return BEFORE;
	}
	Term tmp = car;
	int rc;
	for (int i=0; i<2; i++) {
	    rc = tmp.compareTo(args[i].dereference());
	    if (rc != EQUAL) 
		return rc;
	    tmp = cdr();
	}
	return EQUAL;
    }

	@Override
	public final boolean isImmutable() {
		return immutable;
	}

	/**
	 * adds given term to the end of the list.
	 * Default implementation recreates the whole list and returns reference to new list,
	 * because the original list can be immutable.
	 * */
	public ListTerm add(Term term) {
		Deque<Term> stack = new ArrayDeque<Term>();
		Term t = this;
		while (t instanceof ListTerm){
			ListTerm lt = (ListTerm) t;
			stack.push(lt.car.dereference());
			t = lt.cdr().dereference();
		}
		t = new ListTerm(term, Prolog.Nil);
		while (!stack.isEmpty()){
			t = new ListTerm(stack.pop(), t);
		}
		return (ListTerm) t;
	}

}

File: src/machine/SxxMachine/ListTerm.cpp
using namespace std;

#include "ListTerm.h"
#include "Term.h"
#include "Trail.h"
#include "Prolog.h"
#include "SymbolTerm.h"
#include "StringBuilder.h"

namespace SxxMachine
{

	Term* ListTerm::ArgNoDeRef(const int& i)
	{
		return argz[i];
	}

	bool ListTerm::isCons()
	{
		return true;
	}

	bool ListTerm::isAtomicValue()
	{
		return false;
	}

	int ListTerm::containsTermImpl(Term* variableTerm, Comparator* comparison)
	{
		return car()->containsTerm(variableTerm, comparison) + cdr()->containsTerm(variableTerm, comparison);
	}

	ListTerm::ListTerm()
	{
		//super(s);
	}

	ListTerm::ListTerm(Term* _car, Term* _cdr)
	{
	//	super(".");
		//		// TODO assert _car!=null;
		//    	// TODO assert _cdr!=null;
		//    	//this.car = _car;
		//		//this.cdr = _cdr;
		argz = VA({ _car, _cdr });
		//		//this.immutable = this.argz[0].isImmutable() && this.argz[1].isImmutable();
	}

	Term* ListTerm::car()
	{
		return this->argz[0];
	}

	Term* ListTerm::cdr()
	{
		return this->argz[1];
	}

	bool ListTerm::unifyImpl(Term* t, Trail* trail)
	{
		Term* p = this;
		t = t->dref();
		while(t->isCons() && p->isCons() && p->car()->unify(t->car(), trail))
		{
			p = p->cdr()->dref();
			t = t->cdr()->dref();
		}
		if(t->isVar())
		{
			return t->bind(p, trail);
		}
		if(p->isVar())
		{
			return p->bind(t, trail);
		}
		return !(t->isCons()) && !(p->isCons()) && p->unify(t, trail);
	}

	void ListTerm::setCar(Term* t)
	{
		this->argz[0] = t;
	}

	void ListTerm::setCdr(Term* t)
	{
		this->argz[1] = t;
	}

	int ListTerm::type()
	{
		return TYPE_LIST;
	}

	bool ListTerm::convertible(type_info type)
	{
		return convertible(vector::typeid, type);
	}

	Term* ListTerm::copyImpl(IdentityHashMap<any, Term*>* copyHash, const int& deepCopy)
	{
		if(this->isImmutable())
		{
			return this;
		}
		Deque<ListTerm*>* stack = new ArrayDeque<ListTerm*>();
		Term* p = this;
		while(p->isCons() && !p->asListTerm()->immutable)
		{
			ListTerm* lt = static_cast<ListTerm*>(p);
			stack->push(lt);
			p = lt->cdr()->dref();
		}
		p = p->copy(copyHash, deepCopy);
		while(!stack->isEmpty())
		{
			ListTerm* lt = stack->pop();
			p = CONS(lt->car()->copy(copyHash, deepCopy), p);
		}
		return p;
	}

	bool ListTerm::isGround()
	{
		return this->argz[0]->isGround() && cdr()->isGround();
	}

	int ListTerm::arity()
	{
		return 2;
	}

	std::vector<Term*> ListTerm::args()
	{
		return argz;
	}

	Term* ListTerm::functor()
	{
		return Prolog::SYM_DOT;
	}

	wstring ListTerm::name()
	{
		return Prolog::SYM_DOT->name();
	}

	Term* ListTerm::arg0(const int& nth)
	{
		return nth0(nth);
	}

	Term* ListTerm::nth0(const int& nth)
	{
		Term* t = this;
		int old_nth = nth;
		while((t->isCons()) && 0 < nth)
		{
			nth--;
			t = (t)->cdr()->dref();
		}
		if((t->isCons()))
		{
			return t->car();
		}
		throw out_of_range(old_nth);
	}

	int ListTerm::length()
	{
		int count = 0;
		Term* t = this;
		while((t->isCons()))
		{
			count++;
			t = (t)->cdr()->dref();
		}
		return count;
	}

	any ListTerm::toJava()
	{
		vector<any> vec = vector<any>();
		Term* t = this;
		while((t->isCons()))
		{
			vec.push_back((t)->car()->dref()->toJava());
			t = (t)->cdr()->dref();
		}
		if(!t->isNil())
		{
			vec.push_back(t);
		}
		return vec;
	}

	void ListTerm::toQuotedString_old(const int& printFlags, StringBuilder* sb)
	{
		TermTreeIterator* it = new TermTreeIterator(this, true);
		while(it->hasNext())
		{
			Term* t = it->next();
			t->toQuotedString(0, sb);
			it++;
		}
	}

	void ListTerm::toStringImpl(const int& printingFlags, StringBuilder* sb)
	{
		Term* x = this;
		sb->append("[");
		for(;;)
		{
			x->car()->dref()->toQuotedString(1, sb);
			x = x->cdr()->dref();
			if(!(x->isCons()))
			{
				break;
			}
			sb->append(",");
		}
		if(!Prolog::Nil->equals(x))
		{
			sb->append("|");
			x->toQuotedString(printingFlags, sb);
		}
		sb->append("]");
	}

	bool ListTerm::equalsTerm(Term* obj, Comparator* comparator)
	{
		return obj->isCons() && this->argz[0]->equalsTerm((obj)->car()->dref(), comparator) && cdr()->equalsTerm((obj)->cdr()->dref(), comparator);
	}

	int ListTerm::termHashCodeImpl()
	{
		int h = 1;
		h = 31 * h + Prolog::SYM_DOT->termHashCode();
		h = 31 * h + this->car()->dref()->termHashCode();
		h = 31 * h + cdr()->dref()->termHashCode();
		return h;
	}

	Iterator<Term*>* ListTerm::iterator(const bool& includeSyntax)
	{
		return new ListTermIterator(this, includeSyntax);
	}

SymbolTerm* const  ListTerm::ListTermIterator::LEFT_BRACKET = SymbolTerm::internToken("[");
SymbolTerm* const  ListTerm::ListTermIterator::SEPARATOR = SymbolTerm::internToken("|");
SymbolTerm* const  ListTerm::ListTermIterator::RIGHT_BRACKET = SymbolTerm::internToken("]");
SymbolTerm* const  ListTerm::ListTermIterator::COMMA = SymbolTerm::internToken(",");

	ListTerm::ListTermIterator::ListTermIterator(Term* start, const bool& includeSyntax)
	{
		this->includeSyntax = includeSyntax;
		this->current = start;
	}

	bool ListTerm::ListTermIterator::hasNext()
	{
		return this->current != nullptr && (this->tail.empty() || this->index < this->tail.size());
	}

	Term* ListTerm::ListTermIterator::next()
	{
		Term* result;
		if(this->first && includeSyntax)
		{
			this->first = false;
			return LEFT_BRACKET;
		} else if(this->comma && includeSyntax)
		{
			this->comma = false;
			return COMMA;
		} else if(this->current->isCons())
		{
			result = (this->current)->car();
			this->current = (this->current)->cdr()->dref();
			this->comma = (this->current->isCons());
		} else if(this->current->isNil() && includeSyntax)
		{
			result = RIGHT_BRACKET;
			this->current = nullptr;
		} else if(this->tail.empty())
		{
			this->tail = std::vector<Term*> { this->current, RIGHT_BRACKET };
			result = SEPARATOR;
			this->index = 0;
		} else if(this->index < this->tail.size())
		{
			result = this->tail[this->index];
			this->index++;
		} else
		{
			throw NoSuchElementException();
		}
		return result;
	}

	int ListTerm::compareTo(Term* otherterm)
	{ // otherterm must be dereferenced.
		if((otherterm->isVar()) || (otherterm->isNumber()) || (otherterm->isSymbol()))
		{
			return AFTER;
		}
		if((otherterm->isStructure()))
		{
			int arity = otherterm->arity();
			if(2 != arity)
			{
				return (2 - arity);
			}
			Term* functor = (otherterm)->functor();
			if(!Prolog::SYM_DOT->equalsTerm(functor, StrictEquals))
			{
				return Prolog::SYM_DOT->compareTo(functor);
			}
		}
		std::vector<Term*> args(2);
		if((otherterm->isCons()))
		{
			args[0] = (otherterm)->car();
			args[1] = (otherterm)->cdr();
		} else if((otherterm->isStructure()))
		{
			args = (otherterm)->args();
		} else
		{
			return BEFORE;
		}
		Term* tmp = this->argz[0];
		int rc;
		for(int i = 0; i < 2; i++)
		{
			rc = tmp->compareTo(args[i]->dref());
			if(rc != EQUAL)
			{
				return rc;
			}
			tmp = cdr();
		}
		return EQUAL;
	}

	bool ListTerm::isImmutable()
	{
		return this->immutable;
	}

	ListTerm* ListTerm::add(Term* term)
	{
		if(isImmutable())
		{
			return addToCopy(term);
		}
		Term * const cdr = argz[1];
		if(cdr->isCons())
		{
			return cdr->add(term);
		}
		if(cdr == Prolog::Nil)
		{
			// proper list
			ListTerm* acdr = CONS(term, cdr);
			argz[1] = acdr;

		} else
		{
			// improper list?
			ListTerm* acdr = CONS(term, cdr);
			argz[1] = acdr;
		}
		return this;
	}

	ListTerm* ListTerm::addToCopy(Term* term)
	{
//JAVA TO C++ CONVERTER TODO TASK: Most Java annotations will not have direct C++ equivalents:
//ORIGINAL LINE: @SuppressWarnings("unused") java.util.Deque<Term> stack = new java.util.ArrayDeque<Term>();
		Deque<Term*>* stack = new ArrayDeque<Term*>();
		Term* t = this;
		while(t->isCons())
		{
			ListTerm* lt = static_cast<ListTerm*>(t);
			stack->push(lt->argz[0]->dref());
			t = lt->cdr()->dref();
		}
		t = term->isNil() ? term : CONS(term, Prolog::Nil);
		while(!stack->isEmpty())
		{
			t = CONS(stack->pop(), t);
		}
		return static_cast<ListTerm*>(t);
	}

	ListTerm* ListTerm::append(Term* term)
	{
		if(isImmutable())
		{
			throw NoSuchElementException("isImmutable: " + this);
			//return addToCopy(term);
		}
		Term * const cdr = argz[1];
		if(cdr->isCons())
		{
			return cdr->add(term);
		}
		if(cdr == Prolog::Nil)
		{
			// proper list
			ListTerm* acdr = CONS(term, cdr);
			argz[1] = acdr;
			return acdr;
		} else
		{
			// improper list?
			ListTerm* acdr = CONS(cdr, term);
			argz[1] = acdr;
			return acdr;
		}
	}
}
